src/clj/rems/db/applications.clj

```diff
@@ -52,256 +305,88 @@
-   :blacklisted? #(cache/lookup-or-miss blacklist-cache [%1 %2] (fn [[userid resource]]
-                                                                  (blacklist/blacklisted? userid resource)))
-   ;; TODO: no caching for these, but they're only used by command handlers currently
-   :get-attachment-metadata attachments/get-attachment-metadata
-   :get-catalogue-item-licenses get-catalogue-item-licenses})
-
-(defn get-application-internal
-  "Returns the full application state without any user permission
-   checks and filtering of sensitive information. Don't expose via APIs."
-  [application-id]
-  (let [events (events/get-application-events application-id)]
-    (if (empty? events)
-      nil ; application not found
-      (model/build-application-view events fetcher-injections))))
-
-(defn get-application
-  "Full application state with internal information hidden. Not personalized for any users. Suitable for public APIs."
-  [application-id]
-  (when-let [application (get-application-internal application-id)]
-    (-> application
-        (model/hide-non-public-information)
-        (model/apply-privacy-by-roles #{:reporter})))) ;; to populate required :field/private attributes
-
-(defn get-application-for-user
-  "Returns the part of application state which the specified user
-   is allowed to see. Suitable for returning from public APIs as-is."
-  [user-id application-id]
-  (when-let [application (get-application-internal application-id)]
-    (or (model/apply-user-permissions application user-id)
-        (throw-forbidden))))
+(defmacro one-at-a-time! [& body]
+  ;; Use locks to prevent multiple commands being executed in parallel.
+  ;; Serializable isolation level will already avoid anomalies, but produces
+  ;; lots of transaction conflicts when there is contention. This lock
+  ;; roughly doubles the throughput for rems.db.test-transactions tests.
+  ;;
+  ;; To clarify: without this lock our API calls would sometimes fail
+  ;; with transaction conflicts due to the serializable isolation
+  ;; level. The transaction conflict exceptions aren't handled
+  ;; currently and would cause API calls to fail with HTTP status 500.
+  ;; With this lock, API calls block more but never result in
+  ;; transaction conflicts.
+  ;;
+  ;; See docs/architecture/010-transactions.md for more info.
+  `(try
+     (jdbc/execute! db/*db* ["LOCK TABLE application_event IN SHARE ROW EXCLUSIVE MODE"])
+     ~@body
+
+     (catch org.postgresql.util.PSQLException e#
+       (if (.contains (.getMessage e#) "lock timeout")
+         (throw (TryAgainException. e#))
+         (throw e#)))))
+
+(defn command! [cmd application command-injections]
+  (let [result (commands/handle-command cmd application command-injections)]
+    (when (empty? (:errors result)) ; nothing can't change when there's an error
+      (dependencies/notify-watchers! {:application/id (mapv :application/id (:events result))}))
```

Comment:
Is this double notification? Wouldn't the lower level functions be enough (`add-event` etc.)?